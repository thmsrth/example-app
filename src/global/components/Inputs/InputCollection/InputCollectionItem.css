.InputCollectionItem {
  width: 100%;
  display: flex;
  align-items: center;
  border: 1px solid var(--border-color);
  margin-bottom: var(--section-margin);
}

.InputCollectionItem__delete {
  color: var(--action-secondary-color);
  background-color: var(--action-secondary-background-color);
  border: none;
  font-size: 1.3em;
  padding: 0 var(--section-padding);
  margin-top: calc(var(--section-margin) * 1.5);
}

.InputCollectionItem__delete__icon {
  margin: 0;
}

.InputCollectionItem--alert .InputCollectionItem__delete {
  margin-top: calc(var(--section-margin) * 3);
}

.InputCollectionItem--alert .FormInput__input {
  margin-top: calc(var(--section-margin) * 2);
}

.InputCollectionItemFields {
  flex-grow: 1;
  display: flex;
  flex-wrap: wrap;
  align-items: center;
}

.InputCollectionField {
  flex-grow: 1;
}
