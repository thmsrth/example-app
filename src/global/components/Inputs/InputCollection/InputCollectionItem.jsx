import React from 'react';

import PropTypes from 'prop-types';
import fieldType from './types/field';

import { bemCls } from '../../../utils/ClassNameHelpers';
import { getValidationKeys } from '../../../utils/FormHelpers';

import Button from '../../Button';

import InputCollectionFieldWrapper from './InputCollectionFieldWrapper';

import './InputCollectionItem.css';

const InputCollectionItem = ({
  className, name, value, index, disabled, fields, handleDeleteCollectionItem, ...otherProps
}) => {
  const validationAlert = fields.some(field => (
    getValidationKeys(field, value[index][field.name]).validationAlert
  ));

  return (
    <div className={`${bemCls(`InputCollection FormInput ${className}`, 'Item')} ${
      bemCls(`InputCollection FormInput ${className}`, 'Item--alert', validationAlert)}`}
    >
      <div className={bemCls(`InputCollection FormInput ${className}`, 'ItemFields')}>
        {fields.map(field => (
          <InputCollectionFieldWrapper
            {...otherProps}
            // eslint-disable-next-line react/no-array-index-key
            key={`${name}_${index}_${field.name}`}
            name={name}
            value={value}
            index={index}
            field={field}
            disabled={disabled}
            className={`InputCollectionItem ${className}`}
          />
        ))}
      </div>
      {!disabled
        && (
          <Button
            className={bemCls(`InputCollection FormInput ${className}`, 'Item__delete')}
            text=""
            iconName="remove"
            params={{ index }}
            onClick={handleDeleteCollectionItem}
          />
        )
      }
    </div>
  );
};

export default InputCollectionItem;

InputCollectionItem.propTypes = {
  className: PropTypes.string,
  name: PropTypes.string.isRequired,
  value: PropTypes.arrayOf(PropTypes.objectOf(PropTypes.objectOf(PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.bool,
  ])))),
  disabled: PropTypes.bool,
  index: PropTypes.number.isRequired,
  fields: PropTypes.arrayOf(fieldType),
  handleAddCollectionItem: PropTypes.func,
  handleDeleteCollectionItem: PropTypes.func,
};

InputCollectionItem.defaultProps = {
  className: '',
  value: [],
  fields: [],
  disabled: false,
  handleAddCollectionItem: () => {},
  handleDeleteCollectionItem: () => {},
};
