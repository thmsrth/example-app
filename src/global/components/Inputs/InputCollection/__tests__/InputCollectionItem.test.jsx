import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import InputCollectionItem from '../InputCollectionItem';

const value = [
  {
    myFirstField: {
      isValid: true,
      showWarning: false,
      value: 'http://someurl.foo',
    },
    mySecondField: {
      isValid: true,
      showWarning: false,
      value: 'firstOption',
    },
  },
  {
    myFirstField: {
      isValid: true,
      showWarning: false,
      value: 'http://someotherurl.bar',
    },
    mySecondField: {
      isValid: true,
      showWarning: false,
      value: 'secondOption',
    },
  },
];

const valueWithWarning = [
  {
    myFirstField: {
      isValid: true,
      showWarning: true,
      value: 'http://someurl.foo',
    },
    mySecondField: {
      isValid: true,
      showWarning: true,
      value: 'firstOption',
    },
  },
  {
    myFirstField: {
      isValid: true,
      showWarning: true,
      value: 'http://someotherurl.bar',
    },
    mySecondField: {
      isValid: true,
      showWarning: true,
      value: 'secondOption',
    },
  },
];

const valueWithValidationAlert = [
  {
    myFirstField: {
      isValid: false,
      showWarning: false,
      value: 'http://someurl.foo',
    },
    mySecondField: {
      isValid: false,
      showWarning: false,
      value: 'firstOption',
    },
  },
  {
    myFirstField: {
      isValid: false,
      showWarning: false,
      value: 'http://someotherurl.bar',
    },
    mySecondField: {
      isValid: false,
      showWarning: false,
      value: 'secondOption',
    },
  },
];

const props = {
  className: 'MyInputCollectionItem',
  name: 'MyInputName',
  index: 0,
  fields: [
    {
      name: 'myFirstField',
      title: 'My First Field',
      type: 'text',
      required: true,
      validationAlert: 'Please enter a myFirstField',
      warningAlert: 'My First Field Warning',
    },
    {
      name: 'mySecondField',
      title: 'My Second Field',
      type: 'select',
      required: true,
      validationAlert: 'Please select a mySecondField',
      warningAlert: 'My Second Field Warning',
      options: [
        {
          label: 'First Option',
          value: 'firstOption',
        },
        {
          label: 'Second Option',
          value: 'secondOption',
        },
      ],
    },
  ],
  value,
};

const propsWithWarning = {
  ...props,
  value: valueWithWarning,
};

const propsWithValidationAlert = {
  ...props,
  value: valueWithValidationAlert,
};

describe('InputCollectionItem', () => {
  it('should render the component correctly', () => {
    const tree = shallow(<InputCollectionItem {...props} />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should render the component correctly with warnings for each items field', () => {
    const tree = shallow(<InputCollectionItem {...propsWithWarning} />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should render the component correctly with validation alert for each items field', () => {
    const tree = shallow(<InputCollectionItem {...propsWithValidationAlert} />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});
