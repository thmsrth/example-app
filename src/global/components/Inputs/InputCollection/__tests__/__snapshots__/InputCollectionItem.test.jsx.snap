// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputCollectionItem should render the component correctly 1`] = `
<div
  className="InputCollectionItem FormInputItem MyInputCollectionItemItem "
>
  <div
    className="InputCollectionItemFields FormInputItemFields MyInputCollectionItemItemFields"
  >
    <InputCollectionFieldWrapper
      className="InputCollectionItem MyInputCollectionItem"
      disabled={false}
      field={
        Object {
          "name": "myFirstField",
          "required": true,
          "title": "My First Field",
          "type": "text",
          "validationAlert": "Please enter a myFirstField",
          "warningAlert": "My First Field Warning",
        }
      }
      handleAddCollectionItem={[Function]}
      hidden={false}
      index={0}
      key="MyInputName_0_myFirstField"
      name="MyInputName"
      onBlur={[Function]}
      onChange={[Function]}
      value={
        Array [
          Object {
            "myFirstField": Object {
              "isValid": true,
              "showWarning": false,
              "value": "http://someurl.foo",
            },
            "mySecondField": Object {
              "isValid": true,
              "showWarning": false,
              "value": "firstOption",
            },
          },
          Object {
            "myFirstField": Object {
              "isValid": true,
              "showWarning": false,
              "value": "http://someotherurl.bar",
            },
            "mySecondField": Object {
              "isValid": true,
              "showWarning": false,
              "value": "secondOption",
            },
          },
        ]
      }
    />
    <InputCollectionFieldWrapper
      className="InputCollectionItem MyInputCollectionItem"
      disabled={false}
      field={
        Object {
          "name": "mySecondField",
          "options": Array [
            Object {
              "label": "First Option",
              "value": "firstOption",
            },
            Object {
              "label": "Second Option",
              "value": "secondOption",
            },
          ],
          "required": true,
          "title": "My Second Field",
          "type": "select",
          "validationAlert": "Please select a mySecondField",
          "warningAlert": "My Second Field Warning",
        }
      }
      handleAddCollectionItem={[Function]}
      hidden={false}
      index={0}
      key="MyInputName_0_mySecondField"
      name="MyInputName"
      onBlur={[Function]}
      onChange={[Function]}
      value={
        Array [
          Object {
            "myFirstField": Object {
              "isValid": true,
              "showWarning": false,
              "value": "http://someurl.foo",
            },
            "mySecondField": Object {
              "isValid": true,
              "showWarning": false,
              "value": "firstOption",
            },
          },
          Object {
            "myFirstField": Object {
              "isValid": true,
              "showWarning": false,
              "value": "http://someotherurl.bar",
            },
            "mySecondField": Object {
              "isValid": true,
              "showWarning": false,
              "value": "secondOption",
            },
          },
        ]
      }
    />
  </div>
  <ButtonWrapper
    className="InputCollectionItem__delete FormInputItem__delete MyInputCollectionItemItem__delete"
    iconName="remove"
    onClick={[Function]}
    params={
      Object {
        "index": 0,
      }
    }
    text=""
  />
</div>
`;

exports[`InputCollectionItem should render the component correctly with validation alert for each items field 1`] = `
<div
  className="InputCollectionItem FormInputItem MyInputCollectionItemItem InputCollectionItem--alert FormInputItem--alert MyInputCollectionItemItem--alert"
>
  <div
    className="InputCollectionItemFields FormInputItemFields MyInputCollectionItemItemFields"
  >
    <InputCollectionFieldWrapper
      className="InputCollectionItem MyInputCollectionItem"
      disabled={false}
      field={
        Object {
          "name": "myFirstField",
          "required": true,
          "title": "My First Field",
          "type": "text",
          "validationAlert": "Please enter a myFirstField",
          "warningAlert": "My First Field Warning",
        }
      }
      handleAddCollectionItem={[Function]}
      hidden={false}
      index={0}
      key="MyInputName_0_myFirstField"
      name="MyInputName"
      onBlur={[Function]}
      onChange={[Function]}
      value={
        Array [
          Object {
            "myFirstField": Object {
              "isValid": false,
              "showWarning": false,
              "value": "http://someurl.foo",
            },
            "mySecondField": Object {
              "isValid": false,
              "showWarning": false,
              "value": "firstOption",
            },
          },
          Object {
            "myFirstField": Object {
              "isValid": false,
              "showWarning": false,
              "value": "http://someotherurl.bar",
            },
            "mySecondField": Object {
              "isValid": false,
              "showWarning": false,
              "value": "secondOption",
            },
          },
        ]
      }
    />
    <InputCollectionFieldWrapper
      className="InputCollectionItem MyInputCollectionItem"
      disabled={false}
      field={
        Object {
          "name": "mySecondField",
          "options": Array [
            Object {
              "label": "First Option",
              "value": "firstOption",
            },
            Object {
              "label": "Second Option",
              "value": "secondOption",
            },
          ],
          "required": true,
          "title": "My Second Field",
          "type": "select",
          "validationAlert": "Please select a mySecondField",
          "warningAlert": "My Second Field Warning",
        }
      }
      handleAddCollectionItem={[Function]}
      hidden={false}
      index={0}
      key="MyInputName_0_mySecondField"
      name="MyInputName"
      onBlur={[Function]}
      onChange={[Function]}
      value={
        Array [
          Object {
            "myFirstField": Object {
              "isValid": false,
              "showWarning": false,
              "value": "http://someurl.foo",
            },
            "mySecondField": Object {
              "isValid": false,
              "showWarning": false,
              "value": "firstOption",
            },
          },
          Object {
            "myFirstField": Object {
              "isValid": false,
              "showWarning": false,
              "value": "http://someotherurl.bar",
            },
            "mySecondField": Object {
              "isValid": false,
              "showWarning": false,
              "value": "secondOption",
            },
          },
        ]
      }
    />
  </div>
  <ButtonWrapper
    className="InputCollectionItem__delete FormInputItem__delete MyInputCollectionItemItem__delete"
    iconName="remove"
    onClick={[Function]}
    params={
      Object {
        "index": 0,
      }
    }
    text=""
  />
</div>
`;

exports[`InputCollectionItem should render the component correctly with warnings for each items field 1`] = `
<div
  className="InputCollectionItem FormInputItem MyInputCollectionItemItem "
>
  <div
    className="InputCollectionItemFields FormInputItemFields MyInputCollectionItemItemFields"
  >
    <InputCollectionFieldWrapper
      className="InputCollectionItem MyInputCollectionItem"
      disabled={false}
      field={
        Object {
          "name": "myFirstField",
          "required": true,
          "title": "My First Field",
          "type": "text",
          "validationAlert": "Please enter a myFirstField",
          "warningAlert": "My First Field Warning",
        }
      }
      handleAddCollectionItem={[Function]}
      hidden={false}
      index={0}
      key="MyInputName_0_myFirstField"
      name="MyInputName"
      onBlur={[Function]}
      onChange={[Function]}
      value={
        Array [
          Object {
            "myFirstField": Object {
              "isValid": true,
              "showWarning": true,
              "value": "http://someurl.foo",
            },
            "mySecondField": Object {
              "isValid": true,
              "showWarning": true,
              "value": "firstOption",
            },
          },
          Object {
            "myFirstField": Object {
              "isValid": true,
              "showWarning": true,
              "value": "http://someotherurl.bar",
            },
            "mySecondField": Object {
              "isValid": true,
              "showWarning": true,
              "value": "secondOption",
            },
          },
        ]
      }
    />
    <InputCollectionFieldWrapper
      className="InputCollectionItem MyInputCollectionItem"
      disabled={false}
      field={
        Object {
          "name": "mySecondField",
          "options": Array [
            Object {
              "label": "First Option",
              "value": "firstOption",
            },
            Object {
              "label": "Second Option",
              "value": "secondOption",
            },
          ],
          "required": true,
          "title": "My Second Field",
          "type": "select",
          "validationAlert": "Please select a mySecondField",
          "warningAlert": "My Second Field Warning",
        }
      }
      handleAddCollectionItem={[Function]}
      hidden={false}
      index={0}
      key="MyInputName_0_mySecondField"
      name="MyInputName"
      onBlur={[Function]}
      onChange={[Function]}
      value={
        Array [
          Object {
            "myFirstField": Object {
              "isValid": true,
              "showWarning": true,
              "value": "http://someurl.foo",
            },
            "mySecondField": Object {
              "isValid": true,
              "showWarning": true,
              "value": "firstOption",
            },
          },
          Object {
            "myFirstField": Object {
              "isValid": true,
              "showWarning": true,
              "value": "http://someotherurl.bar",
            },
            "mySecondField": Object {
              "isValid": true,
              "showWarning": true,
              "value": "secondOption",
            },
          },
        ]
      }
    />
  </div>
  <ButtonWrapper
    className="InputCollectionItem__delete FormInputItem__delete MyInputCollectionItemItem__delete"
    iconName="remove"
    onClick={[Function]}
    params={
      Object {
        "index": 0,
      }
    }
    text=""
  />
</div>
`;
